src/OmniSharp.MSBuild/ProjectFile/ProjectFileInfo.ProjectData.cs

```diff
@@ -286,41 +287,37 @@ public static ProjectData Create(MSB.Execution.ProjectInstance projectInstance,
-                    {
-                        projectReferences.Add(fullPath);
-
-                        var aliases = projectReferenceItem.GetMetadataValue(MetadataNames.Aliases);
-                        if (!string.IsNullOrEmpty(aliases))
-                        {
-                            projectReferenceAliases[fullPath] = aliases;
-                        }
-                    }
-                }
 
                 var references = ImmutableArray.CreateBuilder<string>();
                 var referenceAliases = ImmutableDictionary.CreateBuilder<string, string>();
                 foreach (var referencePathItem in projectInstance.GetItems(ItemNames.ReferencePath))
                 {
                     var referenceSourceTarget = referencePathItem.GetMetadataValue(MetadataNames.ReferenceSourceTarget);
+                    var aliases = referencePathItem.GetMetadataValue(MetadataNames.Aliases);
 
+                    // If this reference came from a project reference, count it as such. We never want to directly look
+                    // at the ProjectReference items in the project, as those don't always create project references
+                    // if things like OutputItemType or ReferenceOutputAssembly are set. It's also possible that other
+                    // MSBuild logic is adding or removing properties too.
                     if (StringComparer.OrdinalIgnoreCase.Equals(referenceSourceTarget, ItemNames.ProjectReference))
                     {
-                        // If the reference was sourced from a project reference, we have two choices:
-                        //
-                        //   1. If the reference is a C# project reference, we shouldn't add it because it'll just duplicate
-                        //      the project reference.
-                        //   2. If the reference is *not* a C# project reference, we should keep this reference because the
-                        //      project reference was already removed.
-
-                        var originalItemSpec = referencePathItem.GetMetadataValue(MetadataNames.OriginalItemSpec);
-                        if (originalItemSpec.EndsWith(".csproj", StringComparison.OrdinalIgnoreCase))
+                        var projectReferenceOriginalItemSpec = referencePathItem.GetMetadataValue(MetadataNames.ProjectReferenceOriginalItemSpec);
+                        if (projectReferenceOriginalItemSpec.EndsWith(".csproj", StringComparison.OrdinalIgnoreCase))
                         {
+                            // Convert rooted project reference file paths ("/home/user/source/project/../other_project/project.csproj") to
+                            // full paths and unrooted project reference file paths ("./folder/project.csproj") to rooted full paths
+                            // by combining with the project file's path.
+                            var projectReferenceFilePath = Path.IsPathRooted(projectReferenceOriginalItemSpec)
```

Comment:
@JoeRobich Why the IsRooted check? Path.Combine() is documented as:

> However, if an argument other than the first contains a rooted path, any previous path components are ignored, and the returned string begins with that rooted path component.

Reply:
doh